### PR TITLE
Story5

### PR DESCRIPTION
**Pull request by BigDaveCoding**

### CODE REVIEW

File: app/Http/Requests/CreateAnswerRequest.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+class CreateAnswerRequest extends FormRequest
+{
+    public function rules(): array
+    {
+        return [
+            'answer' => 'required|string',
+            'correct' => 'required|boolean',
+            'question_id' => 'required|integer|exists:questions,id',
+            'feedback' => 'nullable|string',
```
Comment (ashleycoles) [inline app/Http/Requests/CreateAnswerRequest.php]: Can we add in some length checks for these fields? Just to make sure they don't exceed the limit in the DB

File: app/Http/Controllers/AnswerApiController.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+use Illuminate\Http\JsonResponse;
+
+class AnswerApiController extends Controller
+{
+    public function create(CreateAnswerRequest $request): JsonResponse
+    {
+        $answer = new Answer;
+        $answer->answer = $request->answer;
+        $answer->correct = $request->correct;
+        $answer->feedback = $request->feedback ?? null;
```
Comment (ashleycoles) [inline app/Http/Controllers/AnswerApiController.php]: No need for ?? null here - $request->feedback should already contain null if it wasn't sent

File: routes/api.php [hunk 3]
```diff
@@ -11,3 +12,5 @@
     route::post('/quizzes', 'create');
 });
 route::post('/questions', [QuestionApiController::class, 'create']);
```
Comment (ashleycoles) [inline routes/api.php]: Not part of this story, but I just noticed there are some duplicate routes here. The get /quizzes is in here twice. 

Also, please either use the controller group for all quiz routes, or get rid of it and use the long form routes like line 8 and 9

File: tests/Feature/AnswerApiControllerTest.php [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            'correct' => 'false',
+            'question_id' => 'wrong',
+            'feedback' => null,
+        ];
+        $quiz = Quiz::factory()->create(['id' => 1]);
+        $question = Question::factory()->create(['id' => 1, 'quiz_id' => $quiz->id]);
+
+        $response = $this->postJson('/api/answers', $answerData);
+        $response->assertInvalid(['answer', 'correct', 'question_id']);
+    }
```
Comment (ashleycoles) [inline tests/Feature/AnswerApiControllerTest.php:46]: The exists vaidation rule on question_id is a particuarly important one, I think it would be worth adding an extra test that sends all valid data, apart from a question id that doesn't exist

File: tests/Feature/AnswerApiControllerTest.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    public function test_answer_api_controller_successful_creation(): void
+    {
+        $answerData = [
+            'answer' => 'answer',
+            'correct' => true,
+            'question_id' => 1,
+            'feedback' => null,
+        ];
+        $quiz = Quiz::factory()->create(['id' => 1]);
+        $question = Question::factory()->create(['id' => 1, 'quiz_id' => $quiz->id]);
```
Comment (ashleycoles) [inline tests/Feature/AnswerApiControllerTest.php]: Doesn't look like you're using this $question variable

File: tests/Feature/AnswerApiControllerTest.php (hunk 4 above)
Comment (ashleycoles) [reply tests/Feature/AnswerApiControllerTest.php:46]: Probably also worth testing the required fields, send an empty array and make sure those rules work

**Review by ashleycoles — CHANGES_REQUESTED** [review]
Just a few comments!

File: app/Http/Requests/CreateAnswerRequest.php (hunk 1 above)
Comment (BigDaveCoding) [reply app/Http/Requests/CreateAnswerRequest.php]: done :)

File: app/Http/Controllers/AnswerApiController.php (hunk 2 above)
Comment (BigDaveCoding) [reply app/Http/Controllers/AnswerApiController.php]: done

File: routes/api.php (hunk 3 above)
Comment (BigDaveCoding) [reply routes/api.php]: done :)

File: tests/Feature/AnswerApiControllerTest.php (hunk 5 above)
Comment (BigDaveCoding) [reply tests/Feature/AnswerApiControllerTest.php]: It needs a question for the answers to be part of for the one to many relationship. without the question in the database, the test fails.

File: tests/Feature/AnswerApiControllerTest.php (hunk 4 above)
Comment (BigDaveCoding) [reply tests/Feature/AnswerApiControllerTest.php:46]: done

File: tests/Feature/AnswerApiControllerTest.php (hunk 5 above)
Comment (ashleycoles) [reply tests/Feature/AnswerApiControllerTest.php]: I just mean storing the question in $question - you need to use the factory for sure, but it can just be:

`Question::factory()->create(['id' => 1, 'quiz_id' => $quiz->id]);`

File: tests/Feature/AnswerApiControllerTest.php [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+        $response = $this->postJson('/api/answers', $answerData);
+        $response->assertStatus(422)
+            ->assertJson(function (AssertableJson $response) {
+                $response->hasAll('message', 'errors')
+                    ->has('errors', function (AssertableJson $errors) {
+                        $errors->has('question_id')
+                            ->whereType('question_id', 'array');
+                    });
+            });
```
Comment (ashleycoles) [inline tests/Feature/AnswerApiControllerTest.php]: Just use assertInvalid instead

File: tests/Feature/AnswerApiControllerTest.php [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                $response->hasAll('message', 'errors')
+                    ->has('errors', function (AssertableJson $errors) {
+                        $errors->hasAll('answer', 'correct', 'question_id')
+                            ->whereAllType([
+                                'answer' => 'array',
+                                'correct' => 'array',
+                                'question_id' => 'array',
+                            ]);
+                    });
+            });
```
Comment (ashleycoles) [inline tests/Feature/AnswerApiControllerTest.php]: Just use assertInvalid instead

**Review by ashleycoles — CHANGES_REQUESTED** [review]

**Review by ashleycoles — APPROVED** [review]
